src/arch/x86_64/gdt.rs

```diff
@@ -11,18 +12,73 @@ use x86_64::structures::tss::TaskStateSegment;
 use x86_64::structures::DescriptorTablePointer;
 use x86_64::{PrivilegeLevel, VirtAddr};
 
+/// TSS with port bitmap, allocated consecutively.
+///
+/// Bit in port_bitmap: 0 indicates accessible, 1 indicated inaccessible.
+#[derive(Clone, Copy)]
+#[repr(C, packed)]
+#[cfg(feature = "ioport_bitmap")]
+pub struct TaskStateSegmentPortBitmap {
+    pub tss: TaskStateSegment,
+    // Follow linux, add one extra element.
+    pub port_bitmap: [u8; 1 + Self::BITMAP_VALID_SIZE],
+}
+
+#[cfg(feature = "ioport_bitmap")]
+impl TaskStateSegmentPortBitmap {
+    pub const BITMAP_VALID_SIZE: usize = 8192;
+    pub const PORT_BITMAP_OFFSET: usize = size_of::<TaskStateSegment>();
+
+    fn new() -> Self {
+        const DENY_ALL: u8 = !0;
+        Self {
+            tss: TaskStateSegment::new(),
+            port_bitmap: [DENY_ALL; 1 + Self::BITMAP_VALID_SIZE]
+        }
+    }
+}
+
+#[cfg(feature = "ioport_bitmap")]
+pub fn gsbase_to_bitmap<'a>(gsbase: u64) -> &'a mut [u8] {
```

Comment:
Since we set `gsbase` and assume it will never be changed,
I think this function can be set to `pub fn ioport_bitmap() -> &'static mut [u8]`.

Reply:
Great.